baseclasses/utils/solverHistory.py

```diff
@@ -0,0 +1,354 @@
+"""
+==============================================================================
+BaseClasses: Solver History Class
+==============================================================================
+
+@Description : A general solver history class for storing values from a nonlinear solution.
+"""
+
+# ==============================================================================
+# Standard Python modules
+# ==============================================================================
+from collections import OrderedDict
+import copy
+import os
+import time
+from typing import Optional, Type, Dict, Any, List, Iterable, Union
+import pickle
+
+
+class SolverHistory(object):
+    """The SolverHistory class can be used to store and print various useful values during the execution of a solver.
+
+    NOTE: The implementation of this class contains no consideration of parallelism. If you are using a solverHistory
+    object in your parallel solver, you will need to take care over which procs you make calls to the SolverHistory
+    object on.
+    """
+
+    __slots__ = [
+        "variables",
+        "data",
+        "metadata",
+        "iter",
+        "startTime",
+        "defaultFormat",
+        "testValues",
+        "DEFAULT_OTHER_FORMAT",
+        "DEFAULT_OTHER_VALUE",
```

Comment:
Do we want to label these protected, i.e., prepend these with "_" as the user should not really mess with these ever.

Reply:
Sure, good idea